module-services/service-eink/ServiceEink.cpp

```diff
@@ -189,12 +237,77 @@ namespace service::eink
+        }
+        else {
+            // Each bounding box cover the whole width of the context. They are disjoint and sorted by y coordinate.
+            const auto diffBoundingBoxes = gui::Context::linesDiffs(ctx, *previousContext);
+            if (!diffBoundingBoxes.empty()) {
+                const std::uint16_t gapThreshold = ctx.getH() / 4;
+                const std::uint16_t alignment    = 8;
+
+                const auto mergedBoxes = mergeBoundingBoxes(diffBoundingBoxes, gapThreshold);
+                updateFrames           = makeAlignedFrames(mergedBoxes, alignment);
+
+                previousContext->insert(0, 0, ctx);
+            }
         }
 
+        const auto refreshMode = translateToEinkRefreshMode(message->getRefreshMode());
+
+        if (!updateFrames.empty() || refreshMode > previousRefreshMode) {
+            hal::eink::EinkFrame refreshFrame;
+            if (refreshMode > previousRefreshMode) {
+                refreshFrame = {0, 0, {ctx.getW(), ctx.getH()}};
+                updateFrames = {refreshFrame};
+            }
+            else {
+                refreshFrame = updateFrames.front();
+                refreshFrame.size.height =
+                    updateFrames.back().pos_y + updateFrames.back().size.height - updateFrames.front().pos_y;
+            }
+
+            eInkSentinel->HoldMinimumFrequency();
+
+            const auto status = display->showImage(updateFrames, refreshFrame, ctx.getData(), refreshMode);
+            if (status != hal::eink::EinkStatus::EinkOK) {
+                previousContext.reset();
+                previousRefreshMode = hal::eink::EinkRefreshMode::REFRESH_NONE;
+                LOG_ERROR("Error during drawing image on eink: %s", magic_enum::enum_name(status).data());
+            }
+
+            // TEST
+            // {
+            //     std::stringstream ss;
+            //     // for (auto const &bb : diffBoundingBoxes)
+            //     //     ss << bb.str() << ' ';
+            //     // LOG_INFO("Boxes: %s", ss.str().c_str());
+            //     // ss.str("");
+            //     // for (auto const &bb : mergedBoxes)
+            //     //     ss << bb.str() << ' ';
+            //     // LOG_INFO("Merged boxes: %s", ss.str().c_str());
+            //     // ss.str("");
+            //     for (auto const &f : updateFrames)
+            //         ss << '{' << f.pos_x << ' ' << f.pos_y << ' ' << f.size.width << ' ' << f.size.height << "} ";
+            //     LOG_INFO("Frames: %s", ss.str().c_str());
+            //     ss.str("");
+            //     ss << '{' << refreshFrame.pos_x << ' ' << refreshFrame.pos_y << ' ' << refreshFrame.size.width << ' '
+            //     << refreshFrame.size.height << "} ";
+            //     LOG_INFO("Refresh frame: %s", ss.str().c_str());
+            // }
+        }
+
+        // LOG_INFO("Refresh: %d of context: %d %d", (int)refreshMode, (int)ctx.getW(), (int)ctx.getH());
```

Comment:
If it is not needed, delete it

Reply:
It still is. E.g. there are some problems with Harmony which I have to address. What do you think about putting them in an `#ifdef`? Is there a naming convention for macros enabling debug logs?

Reply:
It will be a better option. Take a look at similar solutions in module-utils/log/api/log/debug.hpp

Reply:
Done